### PR TITLE
cargo-tauri.hook: init

### PR DESCRIPTION
**Pull request by getchoo**
## Description of changes

With Tauri apps getting more popular and added to nixpkgs, I believe it's time we share the tedious setup required to build them. This PR introduces a new hook in `cargo-tauri.hook` that builds on top of `rustPlatform.buildRustPackage` (or more acurately, its hooks) to leverage Tauri's bundling feature in creating packages. It is compatible with the two primary targets we share with Tauri -- Linux and macOS -- and meant to be used alongside pre-existing tools for package managers such as npm, pnpm, and yarn

Supersedes #318117

CC @doronbehar for changes made to `yarnConfigHook`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>insulator2</li>
    <li>kiwitalk</li>
    <li>powerdns-admin</li>
    <li>surrealist</li>
    <li>treedome</li>
  </ul>
</details>
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>auto-changelog</li>
    <li>cargo-tauri</li>
    <li>coc-diagnostic</li>
    <li>codefresh</li>
    <li>diagnostic-languageserver</li>
    <li>dotenv-cli</li>
    <li>get-graphql-schema</li>
    <li>gitbutler</li>
    <li>koodo-reader</li>
    <li>kuro</li>
    <li>listmonk</li>
    <li>modrinth-app</li>
    <li>modrinth-app-unwrapped</li>
    <li>nixpkgs-manual</li>
    <li>postlight-parser</li>
    <li>pot</li>
    <li>screego</li>
    <li>spectral-language-server</li>
    <li>textlint-rule-prh</li>
    <li>vim-language-server</li>
    <li>vimPlugins.coc-diagnostic</li>
    <li>xplorer</li>
    <li>yarnConfigHook</li>
    <li>your_spotify</li>
  </ul>
</details>

`powerdns-admin` has been broken since 574c80a110e903c46fe4d6c2df8127b65e8ba1ca

The remaining packages fail with the now infamous

> error[E0282]: type annotations needed for `Box<_>`

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Eveeifyeve** [comment]
And soon bun I am looking to make a hook

**Comment by doronbehar** [comment]
> CC @doronbehar for changes made to `yarnConfigHook`

They seem good. I don't have time unfortunately to review and help with the rest of the PR :/, so I'm unsubscribing myself.

**Comment by getchoo** [comment]
> I will look into documenting this as this will help build Tauri apps on nix

I've made a section for it in the manual already [here](https://github.com/NixOS/nixpkgs/blob/aa6497a787f43d665884708838e4187be2efade3/doc/hooks/tauri.section.md) -- it should work for a vast majority of projects. There's also the [test-app](https://github.com/NixOS/nixpkgs/tree/aa6497a787f43d665884708838e4187be2efade3/pkgs/by-name/ca/cargo-tauri/test-app) showing a real world example with one of the upstream templates

**Comment by happysalada** [comment]
Editorconfig check is failing
Ofborg is also failing on insulator2

Apart from that this looks good.

**Comment by tengkuizdihar** [comment]
because of newer rust version, I got an error when testing treedome. I can make a fix and a quick release, can the new release be in this PR?

**Comment by getchoo** [comment]
> Editorconfig check is failing

Not really sure what to do there as it's failing due to an icon file

> Ofborg is also failing on insulator2

As said in the original comment, this is due to the `time` crate and unrelated to this PR

> because of newer rust version, I got an error when testing treedome. I can make a fix and a quick release, can the new release be in this PR?

Once the update for treedome is merged into master, you can run `nixpkgs-review pr 335751 --package treedome` and it will automatically merge this PR locally and build treedome. I'll probably end up rebasing on master when I make more changes though, so

**Review by philiptaron — DISMISSED** [review]
The editorconfig CI test giving an error is the canary in the coalmine. nixpkgs should have as little vendored code as possible, and it's quite easy to make a repository for this example app.

File: pkgs/by-name/ca/cargo-tauri/test-app/.gitignore
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/test-app/.gitignore:1]: It's been replaced with a basic example from upstream's source

**Comment by Eveeifyeve** [comment]
I will look into using this once I have a bun install/config hook.

**Comment by getchoo** [comment]
Fixed merge conflict with #335795. Treedome should build now

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2411.section.md:393]: Restricting ourselves to just 80 characters makes this kinda hard to read.

**Review by SuperSandro2000 — COMMENTED** [review]
I think it would be cool, if we would extract the unrelated changes and merge them already to make this PR smaller.

**Comment by getchoo** [comment]
>  if we would extract the unrelated changes and merge them already to make this PR smaller

Changes for the Modrinth App are now in https://github.com/NixOS/nixpkgs/pull/336494 and `yarnConfigHook` in https://github.com/NixOS/nixpkgs/pull/338093. I'll make a follow up PR for Treedome when this and the latter are merged since there's no other good way for it to use this hook right now; I'd also prefer the Modrinth App PR be merged prior to this as it'll be a bit easier to rebase on lol

**Comment by Eveeifyeve** [comment]
I am looking to make hooks for bun so if they want to use bun they can which will close this issue #335534

**Comment by happysalada** [comment]
How do we move this forward ?

@philiptaron youre against the adding of test app fo nixpkgs , is that correct ?
I understand this was added as an example for how to use cargohook , or did i misunderstand ? 
Would the path forward be to improve documentation and remove test app ?

I dont have any strong opinons either ways, just trying to brainstorm a path to merging

**Comment by honnip** [comment]
The test app was replaced by an example app on upstream, but that was removed in https://github.com/tauri-apps/tauri/pull/10743

And the location of cargo-tauri has changed (https://github.com/tauri-apps/tauri/pull/10796), so I think it would be good to update it anyway.

**Review by philiptaron — COMMENTED** [review]
I'm out of my depth on the actual content of this PR (the hook creation and use). I _do_ like this PR and hope to learn more.

Here's a review of the content as I read through it this morning.

**Comment by MDM23** [comment]
I'm currently updating `surrealist` to the latest version, which requires cargo-tauri v2. Do you see any way to have v1 and v2 as top-level packages with the new hook?

Initially, I just overwrote it locally:

https://github.com/NixOS/nixpkgs/pull/335619/files#diff-7593db6b9b5b36a086f481cfdce0d8629e002817795794ab40592d321f2e93c2R27

File: doc/hooks/tauri.section.md
Comment (Eveeifyeve) [reply doc/hooks/tauri.section.md:6]: I could have a look into that and add an override to check phase.

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (Eveeifyeve) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:19]: Cargo is needed for the hook to function in hook.sh

File: doc/hooks/tauri.section.md
Comment (philiptaron) [reply doc/hooks/tauri.section.md:6]: I have no idea if it's needed or not! Just, it's sitting between build and install, so it seems likely that if the build phase is different, the test-running phase will be also.

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (philiptaron) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:19]: I like leaving a comment here explaining why things need to get propagated; otherwise, looks good ✅.

File: doc/hooks/tauri.section.md
Comment (getchoo) [reply doc/hooks/tauri.section.md:6]: > I'm a little surprised that there isn't an override for check phase.

AFAIK Tauri doesn't have any official/primary testing tool, only APIs exposed that can be used in basically any library. This makes it pretty hard to make any assumptions about what individual projects will be doing, so I would much rather move that responsibility over to other hooks -- sort of like we do with the Node fetchers/configuration hooks

I think `cargo test` will also cover most projects as a default realistically :p

File: pkgs/by-name/ca/cargo-tauri/test-app.nix
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/test-app.nix:20]: In this case it's actually intentional as the default unpack phase uses the [name](https://github.com/NixOS/nixpkgs/blob/8ed9f95120762d0ed073cd6cfebf41f58347442f/pkgs/stdenv/generic/setup.sh#L1192) of the derivation as the unpack destination in the build dir

The manual [says](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-sourceRoot) for example
> if your want your build to start in a sub-directory inside your sources, and you are using fetchzip-derived src (like fetchFromGitHub or similar), you need to set sourceRoot = "${src.name}/my-sub-directory".

(and yes I've [fixed](https://github.com/NixOS/nixpkgs/pull/339604) the typo 😆)

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (Eveeifyeve) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:19]: Yeah I could look at helping @getchoo by adding a comment.

**Comment by nyabinary** [comment]
Shouldn't this be firxed as well? https://github.com/tauri-apps/tauri/issues/8679

**Comment by getchoo** [comment]
> Shouldn't this be fixed as well? https://github.com/tauri-apps/tauri/issues/8679

I think that would be something more for upstream, as the only way we can "fix" this is by applying the patch linked in the issue to each package. I wouldn't consider that in scope for this PR though, and probably not a good idea to do across the board as messing with vendored deps should generally be avoided. For packages that need the patch, they should be able to use it in [`crateOverrides`](https://github.com/NixOS/nixpkgs/blob/4764ce988357a7f7ce0b4a562145977c5efbc712/doc/languages-frameworks/rust.section.md#handling-external-dependencies-handling-external-dependencies) similar to any other Rust package

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:19]: > Why does this hook need to propagate cargo?

I don't think it really does need to, but this appears to be an [internal implementation detail](https://github.com/NixOS/nixpkgs/blob/9c5c04ca4361effd7e0359dde6dae7c12822f243/pkgs/build-support/trivial-builders/default.nix#L592) of `makeSetupHook`

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:44]: I don't think there should be any cases where this can happen, but yeah I agree this is a big footgun. As the single directory (currently) matched by this glob is named after the application though (which may be different from our package's `pname`, etc), I'm not sure how else to go about this

Maybe we could try to figure out the name from the Tauri configuration?

File: pkgs/by-name/ca/cargo-tauri/hook.sh
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.sh:32]: Yep, `--target-dir` is respected now. What I'm trying to explain here though is that when it's used alongside the target `--target` argument, it still tries to use `target/<release type>` as the target directory even though `cargo` will actually output to `target/<target triple>/<release type>`

I should probably make a patch/issue for this at some point, I just haven't had the time lol

File: pkgs/by-name/ca/cargo-tauri/hook.sh
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.sh:48]: I'd imagine some do. This is from the original cargoBuildHook though, so we should probably keep this behavior

**Comment by philiptaron** [comment]
Looks like #335619 sniped you.

**Comment by Eveeifyeve** [comment]
also we should make this a generic or something as there is dev version of cargo tauri cli.

**Comment by Eveeifyeve** [comment]
And we should start supporting it as it's in rc.

**Comment by getchoo** [comment]
> also we should make this a generic or something as there is dev version of cargo tauri cli.

Not sure how well that would work given the breaking changes now (and possibly in the future) in Tauri v2 from v1. I would consider these hooks to be pretty flexible though, so if there aren't that many breaking changes related to the build I doubt it would be hard to implement

> And we should start supporting it as it's in rc.

That would be best in another PR

**Comment by nyabinary** [comment]
> > also we should make this a generic or something as there is dev version of cargo tauri cli.
> 
> Not sure how well that would work given the breaking changes now (and possibly in the future) in Tauri v2 from v1. I would consider these hooks to be pretty flexible though, so if there aren't that many breaking changes related to the build I doubt it would be hard to implement
> 
> > And we should start supporting it as it's in rc.
> 
> That would be best in another PR

https://v2.tauri.app/blog/roadmap-to-tauri-2-0/

There won't be any breaking changes in RC according to this so I think it's best to do it in this pr tbh

**Comment by getchoo** [comment]
> There won't be any breaking changes in RC

I'm talking about the breaking changes from Tauri v1 -> Tauri v2, not breaking changes from the v2 RC to stable

> so I think it's best to do it in this pr tbh

Setting up the v2 packaging here initially is definitely out of scope IMO given how large this PR already is. The hooks created here will -- at the very least -- require changes to most dependencies like using GTK4 over 3, a newer webkitgtk ABI, etc. I think it would be much better to make those changes later on once we get this merged and actually package v2

**Comment by philiptaron** [comment]
@getchoo -- please make sure to integrate the changes from #340862 when you fix the merge conflict.

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:44]: Now I think about it again: a better option would be to use `jq` on `tauri.conf`. We should reliably be able to find that file based on the current build directory (`buildAndTestSubdir`)

File: pkgs/by-name/ca/cargo-tauri/hook.sh
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.sh:30]: This still doesn't work for the reasons mentioned above

File: pkgs/by-name/ca/cargo-tauri/test-app.nix
Comment (Scrumplex) [inline pkgs/by-name/ca/cargo-tauri/test-app.nix:25]: Could you generate this patch by running `cargo update time` instead? That should reduce the size of the patch I think

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (getchoo) [reply pkgs/by-name/ca/cargo-tauri/hook.nix:44]: TL;DR: After revisiting this, it seems we will need to keep it

Tauri's deb bundles output both a packaged .deb and an unpacked version to `$targetDir/bundle/deb`. The basename of both of these is in the format of `<productName>_<version>_<arch>`, with `<arch>` being determined by [this](https://github.com/tauri-apps/tauri/blob/58bab8b35b6e3d4ddcc229920081c878bd7c7fd6/crates/tauri-bundler/src/bundle/linux/debian.rs#L44-L48) match statement internally. We can grab the `productName` from `tauri.conf.json` (which should always be in the current build directory as said before) and figure out the architecture by reimplementing that same match statement; ***however***, we have no consistent way to find the version as that is read from Cargo.toml, which may have it under `package.version` or `workspace.package.version`. The best we could do is attempt to find it in either of these and fail/fallback to a variable if nothing is found, but IMO this is introducing a lot of complication for something that's still quite fragile

We should be fine though, as the only way for multiple directories to be matched here would be if we're building multiple targets, which this hook doesn't support in the first place

**Comment by Eveeifyeve** [comment]
I think the hook should support different options of bundling as Darwin has app,msg and pkgs and Linux you get the point

**Comment by getchoo** [comment]
The recent pushes should fix all build failures (aside from `insulator2`, but that's from the infamous `time` crate issue) and newly added packages have been updated to use the hooks as well. This should (hopefully) be ready for a final round of review

Result of `nixpkgs-review pr 335751` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>modrinth-app</li>
    <li>modrinth-app-unwrapped</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>insulator2</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>cargo-tauri</li>
    <li>en-croissant</li>
    <li>gitbutler</li>
    <li>kiwitalk</li>
    <li>mouse-actions-gui</li>
    <li>nixpkgs-manual</li>
    <li>pot</li>
    <li>surrealist</li>
  </ul>
</details>

**Comment by getchoo** [comment]
> I think the hook should support different options of bundling as Darwin has app,msg and pkgs and Linux you get the point

The use of bundles here is mainly an implementation detail to use packaging related items generated by Tauri (like desktop files). I don't know what purpose this would really serve here in nixpkgs

**Comment by Eveeifyeve** [comment]
> > I think the hook should support different options of bundling as Darwin has app,msg and pkgs and Linux you get the point
> 
> 
> 
> The use of bundles here is mainly an implementation detail to use packaging related items generated by Tauri (like desktop files). I don't know what purpose this would really serve here in nixpkgs

Well it would have a purpose of giving the developer choice on how they will package their apps.

**Comment by Eveeifyeve** [comment]
> nixpkgs-review pr 335751

I am also testing on `aarch64-darwin`.

**Comment by Eveeifyeve** [comment]
Result of `nixpkgs-review pr 335751` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

<details>
<summary>4 packages marked as broken and skipped:</summary>
<ul>
<li>cinny-desktop</li>
<li>kiwitalk modrinth-app</li>
<li>modrinth-app</li>
<li>modrinth-app-unwrapped</li>
</ul>
</details>
<details>
<summary>2 packages failed to build:</summary>
<ul>
 <li>en-croissant</li>
 <li>gitbutler</li>
 <ul>
 </details>

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (Eveeifyeve) [inline pkgs/by-name/ca/cargo-tauri/hook.nix:41]:
```nix
        darwin = ''
          mkdir $out/Applications
          mv "$targetDir"/bundle/macos $out/Applications
        '';
```

Comment (Eveeifyeve) [inline pkgs/by-name/ca/cargo-tauri/hook.nix:41]:
```nix
        darwin = ''
          mkdir $out/Applications
          mv "$targetDir"/bundle/macos $out/Applications
        '';
```

**Review by Eveeifyeve — COMMENTED** [review]
Perhaps it might fix the cinny desktop build issue?

File: pkgs/by-name/ca/cargo-tauri/hook.nix
Comment (Eveeifyeve) [inline pkgs/by-name/ca/cargo-tauri/hook.nix:41]: To fix an issue with cinny desktop of `$out/Applications` not existing.
```suggestion
        darwin = ''
          mkdir $out/Applications
          mv "$targetDir"/bundle/macos $out/Applications
        '';
```

**Comment by getchoo** [comment]
> Well it would have a purpose of giving the developer choice on how they will package their apps.

This hook isn't meant for developers to package and distribute their apps outside of nixpkgs. Again, this is an implementation detail; there would be no point in outputting an appimage, deb package, etc

If you *really* wanted to do this though, `tauriBundleType` is [available](https://github.com/NixOS/nixpkgs/blob/d90dd95acaec3a98e397ca496778ae0b7e22e3ca/doc/hooks/tauri.section.md#tauribundletype-tauri-bundle-type), but it's primarily meant to be used for platforms not supported by this hook at all/very special cases. It will require you to add your own dependencies and installation script to work in the way you describe (but fair warning: Nix binaries will probably not work outside of nixpkgs)

**Review by philiptaron — APPROVED** [review]
Thanks for your contribution, Seth (@getchoo)!

I'm approving. This looks good to me. I plan on merging in another day or so unless someone in the greater Rust-in-nixpkgs community requests changes. (CC @emilazy?)

The build failure of `insulator2` is part of #332957.

In what I hope is a my-system-issue, both on master and on this PR, I can't run any of the binaries to validate that they're functioning.

For example,

```
** (kiwi-talk:248290): WARNING **: 12:20:51.751: webkit_settings_set_enable_offline_web_application_cache is deprecated and does nothing.

(kiwi-talk:248290): Gtk-CRITICAL **: 12:20:51.860: gtk_widget_get_scale_factor: assertion 'GTK_IS_WIDGET (widget)' failed
Gdk-Message: 12:20:52.015: Error 71 (Protocol error) dispatching to Wayland display.
```

I pretty strongly suspect this is about how I've configured my system, though, so I'm certainly not blocking this PR on investigating that.

Result of `nixpkgs-review pr 335751` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>cinny-desktop</li>
  </ul>
</details>
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>insulator2</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>cargo-tauri</li>
    <li>en-croissant</li>
    <li>gitbutler</li>
    <li>kiwitalk</li>
    <li>modrinth-app</li>
    <li>modrinth-app-unwrapped</li>
    <li>mouse-actions-gui</li>
    <li>nixpkgs-manual</li>
    <li>pot</li>
    <li>surrealist</li>
    <li>treedome</li>
  </ul>
</details>

**Comment by getchoo** [comment]
Rebased to fix merge conflicts with the release notes

**Review by PhilipTaronQ — APPROVED** [review]

**Review by philiptaron — APPROVED** [review]

**Pull request merged**
